modal/_container_entrypoint.py

```diff
@@ -444,27 +487,39 @@ async def call_function_async(
                 obj.__enter__()
 
     try:
-        async for input_id, args, kwargs in function_io_manager.run_inputs_outputs.aio():
+
+        async def run_input(input_id, args, kwargs):
             output_index = SequenceNumber(0)  # mutable number we can increase from the generator loop
-            async with function_io_manager.handle_input_exception.aio(input_id, output_index):
+            started_at = time.time()
+            async with function_io_manager.handle_input_exception.aio(input_id, started_at, output_index):
                 res = fun(*args, **kwargs)
 
                 # TODO(erikbern): any exception below shouldn't be considered a user exception
                 if is_generator:
                     if not inspect.isasyncgen(res):
                         raise InvalidError(f"Async generator function returned value of type {type(res)}")
                     async for value in res:
-                        await function_io_manager.enqueue_generator_value.aio(input_id, output_index.value, value)
+                        await function_io_manager.enqueue_generator_value.aio(
+                            input_id, started_at, output_index.value, value
+                        )
                         output_index.increase()
-                    await function_io_manager.enqueue_generator_eof.aio(input_id, output_index.value)
+                    await function_io_manager.enqueue_generator_eof.aio(input_id, started_at, output_index.value)
                 else:
                     if not inspect.iscoroutine(res) or inspect.isgenerator(res) or inspect.isasyncgen(res):
                         raise InvalidError(
                             f"Async (non-generator) function returned value of type {type(res)}"
                             " You might need to use @stub.function(..., is_generator=True)."
                         )
                     value = await res
-                    await function_io_manager.enqueue_output.aio(input_id, output_index.value, value)
+                    await function_io_manager.enqueue_output.aio(input_id, started_at, output_index.value, value)
+
+        if input_concurrency > 1:
+            async with TaskContext() as execution_context:
```

Comment:
actually i don't think we should use tasks here. `TaskContext` actually has an O(N^2) complexity while terminating iirc. but also asyncio tasks have a bunch of overhead and are annoying to debug. i think you probably want to create more like a pool and then join the coroutines in the end maybe? there's nothing built into asyncio unfortunately

Reply:
this one might be good? https://github.com/gistart/asyncio-pool

otoh for your use case you can also just implement a small utility function and it's probably like 10 lines.

Reply:
Happy to create a small utility class that is a subset of `TaskContext` - but maybe one of these points changes the picture?

- I think `TaskContext.wait` is `O(N^2)` but we don't use that here. Instead, `.stop()` is called on exit with no grace, meaning any in-progress tasks just get cancelled.

- Also, the overhead of coroutines should scale with the value of `allow_concurrent_inputs` set by the user. It would be potentially confusing if the user sets some `n` but only ever sees `<n` executing at one time.

Reply:
aren't you kicking off one task for every input?

Reply:
i think asyncio.Task has some minor overhead over just a straight up coroutine, so if you can use the latter only, that might be preferable

Reply:
Do you think we can group this into the medium-sized refactor, which would involve:

- Refactor `run_inputs_outputs` to no longer be a generator, but instead expose `fetch_input` on `FunctionIOManager` API. `fetch_input` can return an EOF. `fetch_input` holds a lock in `FunctionIOManager` to prevent concurrent requests going out to the worker
- Sync/Async call-loops can now involve continuously spawning `run_inputs` into an asyncio pool, or `ThreadPoolExecutor` to manage the number of concurrent `run_inputs` calls
- `run inputs` involves: acquiring semaphore, calling `fetch_input`, running the user function and finally `enqueue_output`

Reply:
Yeah something like that makes sense!